Updating Figma content to include variables (#653)
primer/design: content/guides/figma/getting-started.mdx

---
title: Getting started
description: Principles, standards, and usage guidelines for designing GitHub interfaces with Figma
---
import {LinkExternalIcon} from '@primer/octicons-react'

> ## What are Primer Figma libraries?

Author: They are not though. They are published to figma community.

> The Primer Figma libraries contain UI components and design tokens (`variables` and `styles`) that our teams at GitHub use to design GitHub. The components contained within Primer match what is available for developers in Primer React Components, Primer ViewComponents, and Primer CSS.

Reviewer: If we refer to libraries as plural I would like to seen a list of which ones exist and what's inside them.

Author: Fix and I only refer to Primer web now. Great catch.

> Figma libraries contain UI components and design tokens (`variables` and `styles`). At GitHub we have one product design system library called `Primer Web` that our teams at GitHub use to design GitHub. The components contained within Primer match what is available for developers in Primer React Components, Primer ViewComponents, and Primer CSS.

Author: ```suggestion
Figma libraries contain UI components and design tokens (`variables` and `styles`). At GitHub we have one product design system library called _Primer Web_ that our teams at GitHub use to design GitHub. The components contained within Primer match what is available for developers in Primer React Components, Primer ViewComponents, and Primer CSS.
```

  />
</ImageBox>

> <ImageBox caption="Enable 'Primer Web' and all other libraries you need in your file.">

Author: ```suggestion
<ImageBox caption="Enable Primer Web and all other libraries you need in your file.">
```

  <img
    width="508"
    alt="Screenshot showing the library overview in figma"
    src="https://github.com/primer/react/assets/813754/675eb97e-479d-47db-bd29-f10c136f586a"
  />
</ImageBox>

### Styles and variables

> Styles and variables are two very similar ways of bringing `design tokens` like [Primer Primitives](https://github.com/primer/primitives) into Figma.

Reviewer: ```suggestion
Styles and variables are two ways to use [Primer Primitives](https://github.com/primer/primitives) (design tokens) in Figma.
```

Our goal is to move everything over to variables, but at the moment only color and size tokens are supported. This is why we still provide text and shadow tokens using styles.

### Understanding styles

> Figma libraries like `Primer Web` provide styles for you to use. In contrast to local styles, styles from a library don't show up in the sidebar. However once you open a selection to choose a `text style` or `shadow` you will see the styles from team libraries as well.

Reviewer 2: Non-blocking nitpick: I'm not sure we want "Primer Web", "text style", or "shadow" in a `<code>` tag. I think this is intended for things like variable names.

Author: Hey, so for Primer Web, yes.

For text and shadow styles I am wondering if this is correct as they are specific Figma things. Similar to a variable or a property. WDYT?

Reviewer 2: I guess that's ok. I personally wouldn't do it that way, but I don't think anybody will be confused 🙂

> Figma libraries like `Primer Web` provide styles for you to use. In contrast to local styles, styles from a library don't show up in the sidebar. However once you open a selection to choose a `text style` or `shadow` you will see the styles from team libraries as well.

Author: ```suggestion
Figma libraries like _Primer Web_ provide styles for you to use. In contrast to local styles, styles from a library don't show up in the sidebar. However, once you open a selection to choose a `text style` or `shadow` you will see the styles from team libraries as well.
```

<ImageBox caption="Text style selector with search for 'body'">
  <img
    width="261"
    alt="Screenshot showing the style selector in figma"
    src="https://github.com/primer/react/assets/813754/4537b229-a663-499d-b0f0-d6be0d9fdadb"
  />
</ImageBox>

### Understanding variables (primitives)
Our Figma libraries use `variables` to represent design tokens. Variables from a library don't show up in the sidebar.
However once you open a selection to choose a color, or size, etc. you will see the variables from team library as well.

To use the `variables` you need to enable the `Primer Web` library in your file.

Available variables:
- color variables
- size variables

To quickly find a `variable` you can use the search box. E.g. use `accent` to bring up all accent colors or `fgColor` to find all text related colors. 

<ImageBox caption="Color variables panel and panel with a search for 'fgCo'">
  <img
    width="509"
    alt="Screenshot showing the color variable panel in figma"
    src="https://github.com/primer/design/assets/813754/34835137-cc62-45ce-9abf-3573089eb278"
  />
</ImageBox>

#### Scope

> Variables in Figma are *"scoped"*, meaning they can only be used in specific situations. Currently Figma offers the following scopes for color variables: `Frame fill`, `Shape fill`, `Text fill` , `Stroke`.

Reviewer 2: Non-blocking nitpick: I'm not sure we need "Frame fill", "Shape fill", etc in `<code>` blocks. Those just refer to Figma tools.

Author: Same as above. I do feel it is nice to somehow highlight that those are actual scope values. But maybe a code block is not correct, and we should use a different visual style?

Out of interest, why do you feel code blocks are problematic here?

Reviewer 2: I think it's just a matter of personal preference. It makes me thinks these are things we have in Primer code. If nobody else gets hung up on these, I think it's totally fine.

Whenever possible use components as they are without _detaching_.

<a href="https://help.figma.com/hc/en-us/articles/360038662654-Guide-to-components-in-Figma">Learn more about Figma components <LinkExternalIcon /></a>

## Using variable themes

Primer Web provides _light mode_ and _dark mode_ using figma `variables`. This means you don't need a plugin to change between modes. Simply select any `frame` that uses `variables` or components with `variables`. Select the mode from the dropdown <svg class="svg" xmlns="http://www.w3.org/2000/svg" width="14" height="12" viewBox="0 0 14 12"><path fill="#000" fill-opacity="1" fill-rule="evenodd" d="m5 1.381-4 2.31v4.618l4 2.31 4-2.31V3.691L5 1.38zm5 1.732L5 .227 0 3.113v5.774l5 2.887 5-2.887V3.113zm3 5.196-5 2.887 1 .578 5-2.887V3.113L9 .227 8 .804l5 2.887v4.618zM6 6c0 .552-.448 1-1 1-.552 0-1-.448-1-1 0-.552.448-1 1-1 .552 0 1 .448 1 1z"></path></svg> in the _layer_ section in the right sidebar

<ImageBox caption="Switching the variable mode changes all nested items">

>   <img
>     width="800"
>     alt="Screen recording of switching variable modes in Figma"
>     src="https://github.com/primer/design/assets/813754/0675a32c-3c24-40a7-a030-4c60e1b0409f"
>   />

Author: ```suggestion
  <CustomVideoPlayer
  width="804"
  loop
  autoplay
  src="https://github.com/primer/design/assets/813754/b4065f62-0786-44f4-b214-501307fbcb5f"
/>
```

Whenever possible use components as they are without *detaching*.

<a href="https://help.figma.com/hc/en-us/articles/360038662654-Guide-to-components-in-Figma">Learn more about Figma components <LinkExternalIcon /></a>

## Using variable themes

Primer Web provides *light mode* and *dark mode* using figma `variables`. This means you don't need a plugin to change between modes. Simply select any `frame` that uses `variables` or components with `variables`. Select the mode from the dropdown <svg class="svg" xmlns="http://www.w3.org/2000/svg" width="14" height="12" viewBox="0 0 14 12"><path fill="#000" fill-opacity="1" fill-rule="evenodd" d="m5 1.381-4 2.31v4.618l4 2.31 4-2.31V3.691L5 1.38zm5 1.732L5 .227 0 3.113v5.774l5 2.887 5-2.887V3.113zm3 5.196-5 2.887 1 .578 5-2.887V3.113L9 .227 8 .804l5 2.887v4.618zM6 6c0 .552-.448 1-1 1-.552 0-1-.448-1-1 0-.552.448-1 1-1 .552 0 1 .448 1 1z"></path></svg> in the *layer* section in the right sidebar

> <ImageBox caption="Switching the variable mode changes all nested items">

Reviewer: I love little snippets. 😍

The GIF's quality isn't great, but it might be hard to fix unless we change it to a video. We should see what the team thinks, but I think we've talked about using videos instead of GIFs because they're easier for everyone to use.

Reviewer 2: +1 for using a video instead of a GIF here

Author: How can I add an autoplaying video?

Reviewer 2: There's a component called `CustomVideoPlayer`. It accepts all the same props as the HTML [<video>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) tag, so you can pass `autoplay`.

Author: Great, thanks, used that now. It does not support `muted` though, right?

  <img
    width="800"
    alt="Screen recording of switching variable modes in Figma"
    src="https://github.com/primer/design/assets/813754/0675a32c-3c24-40a7-a030-4c60e1b0409f"
  />
</ImageBox>

> ### Video introduction to primer web variables and modes / themes (github internal)

Reviewer 2: Isn't all of the information about Primer Web only relevant to GitHub staff? I didn't think Primer Web was published publicly.

Author: It is published publicly.

> ### Video introduction to primer web variables and modes / themes (github internal)

Reviewer 2: ```suggestion
### Video introduction to primer web variables and modes / themes (GitHub staff only)
```

> <a href="https://github.rewatch.com/video/qvryg0cscoowmpyx-figma-primer-web-update-to-figma-variables"><img src="https://thumbnails.rewatch.com/thumbs/qvryg0cscoowmpyx?fm=jpg&p=true&q=100" /></a>

Reviewer: Note: Remember that this is only for internal use. As I mentioned before, we need to make it clear that it's for internal purposes. If we decide to make the Figma library public in the future, maybe we could download the video and host it on a public URL?

Author: It is publicly available forever already. 😉
